File: client/src/utils/chart/structureMath.ts
/** EMA, RSI, swing fractals (1:1 with trading-cursor-models strategies/lib/structure.py). */

export function confirmSwingHigh(
  highs: number[],
  i: number,
  left: number,
  right: number,
): boolean {
  if (i < left || i + right >= highs.length) return false
  const pivot = highs[i]
  for (let j = i - left; j <= i + right; j++) {
    if (j !== i && highs[j] >= pivot) return false
  }
  return true
}

export function confirmSwingLow(
  lows: number[],
  i: number,
  left: number,
  right: number,
): boolean {
  if (i < left || i + right >= lows.length) return false
  const pivot = lows[i]
  for (let j = i - left; j <= i + right; j++) {
    if (j !== i && lows[j] <= pivot) return false
  }
  return true
}

export function ema(values: number[], period: number): (number | null)[] {
  const n = values.length
  const out: (number | null)[] = Array(n).fill(null)
  if (period <= 0 || n < period) return out
  const k = 2 / (period + 1)
  let seed = 0
  for (let i = 0; i < period; i++) seed += values[i]
  seed /= period
  out[period - 1] = seed
  let prev = seed
  for (let i = period; i < n; i++) {
    prev = values[i] * k + prev * (1 - k)
    out[i] = prev
  }
  return out
}

export function rsi(closes: number[], period: number): (number | null)[] {
  const n = closes.length
  const out: (number | null)[] = Array(n).fill(null)
  if (period <= 0 || n <= period) return out

  let gains = 0
  let losses = 0
  for (let i = 1; i <= period; i++) {
    const d = closes[i] - closes[i - 1]
    if (d >= 0) gains += d
    else losses -= d
  }
  let avgGain = gains / period
  let avgLoss = losses / period
  if (avgLoss === 0) out[period] = 100
  else {
    const rs = avgGain / avgLoss
    out[period] = 100 - 100 / (1 + rs)
  }

  for (let i = period + 1; i < n; i++) {
    const d = closes[i] - closes[i - 1]
    const gain = d > 0 ? d : 0
    const loss = d < 0 ? -d : 0
    avgGain = (avgGain * (period - 1) + gain) / period
    avgLoss = (avgLoss * (period - 1) + loss) / period
    if (avgLoss === 0) out[i] = 100
    else {
      const rs = avgGain / avgLoss
      out[i] = 100 - 100 / (1 + rs)
    }
  }
  return out
}

File: src/PolyTrader.Core/Strategy/StructureMath.cs
namespace PolyTrader.Core.Strategy;

/// <summary>EMA, RSI, and swing fractals (1:1 with trading-cursor-models strategies/lib/structure.py).</summary>
public static class StructureMath
{
    public static bool ConfirmSwingHigh(
        IReadOnlyList<double> highs,
        int i,
        int left,
        int right)
    {
        if (i < left || i + right >= highs.Count)
        {
            return false;
        }

        var pivot = highs[i];
        for (var j = i - left; j <= i + right; j++)
        {
            if (j != i && highs[j] >= pivot)
            {
                return false;
            }
        }

        return true;
    }

    public static bool ConfirmSwingLow(
        IReadOnlyList<double> lows,
        int i,
        int left,
        int right)
    {
        if (i < left || i + right >= lows.Count)
        {
            return false;
        }

        var pivot = lows[i];
        for (var j = i - left; j <= i + right; j++)
        {
            if (j != i && lows[j] <= pivot)
            {
                return false;
            }
        }

        return true;
    }

    public static double?[] Ema(IReadOnlyList<double> values, int period)
    {
        var n = values.Count;
        var outVals = new double?[n];
        if (period <= 0 || n < period)
        {
            return outVals;
        }

        var k = 2.0 / (period + 1);
        var seed = values.Take(period).Average();
        outVals[period - 1] = seed;
        var prev = seed;
        for (var i = period; i < n; i++)
        {
            prev = values[i] * k + prev * (1 - k);
            outVals[i] = prev;
        }

        return outVals;
    }

    public static double?[] Rsi(IReadOnlyList<double> closes, int period)
    {
        var n = closes.Count;
        var outVals = new double?[n];
        if (period <= 0 || n <= period)
        {
            return outVals;
        }

        var gains = 0.0;
        var losses = 0.0;
        for (var i = 1; i <= period; i++)
        {
            var d = closes[i] - closes[i - 1];
            if (d >= 0)
            {
                gains += d;
            }
            else
            {
                losses -= d;
            }
        }

        var avgGain = gains / period;
        var avgLoss = losses / period;
        if (avgLoss == 0)
        {
            outVals[period] = 100.0;
        }
        else
        {
            var rs = avgGain / avgLoss;
            outVals[period] = 100.0 - 100.0 / (1.0 + rs);
        }

        for (var i = period + 1; i < n; i++)
        {
            var d = closes[i] - closes[i - 1];
            var gain = d > 0 ? d : 0.0;
            var loss = d < 0 ? -d : 0.0;
            avgGain = (avgGain * (period - 1) + gain) / period;
            avgLoss = (avgLoss * (period - 1) + loss) / period;
            if (avgLoss == 0)
            {
                outVals[i] = 100.0;
            }
            else
            {
                var rs = avgGain / avgLoss;
                outVals[i] = 100.0 - 100.0 / (1.0 + rs);
            }
        }

        return outVals;
    }
}
